### ISSUE TITLE
[Redirector] Facebook on Mobile Data (or false proxy)

### ISSUE THREAD

**Issue by kowith337** [DandelionSprout/adfilt#12]
The majority of using Facebook on mobile browser + mobile data is get ruined by their system that may redirect to unwanted subdomain, such as
- `h.facebook.com` **without proper HTTPS connections**
- `mobile.facebook.com`
- `web.facebook.com`

This also depend one or more conditions about...
- Connect to service via mobile data from the carrier that have `Free Basics` service available
- User agent of browser is pointed/identified that you're using mobile device, e.g. Android, iOS, ....

While I want to...
- Visit `mbasic.facebook.com` to keep using the feature phone layout of Facebook, keep simple, low effects and eye candy.
  > - While use on mobile data, it's redirect to `h.facebook.com` with **NO HTTPS!!!!!!11111**
  > - `mobile.facebook.com` is the final destination of it that depend on your browser's user agent about what site version (between basic and touch) should be loaded.
- Visit `touch.facebook.com` to get (not so) near experience of their bloated app without installing one or bunch of them, and keep using on this subdomain.
  > - You may get same redirect as it happen on `mbasic.facebook.com`, but if you can use some trick to bypass, a seamless transition/navigation become stuck due to it mostly redirect their `AJAX`/`XHR` to `mobile.facebook.com`, rather than stay on same subdomain.
  > - Mostly browsers, even wrapper app have forbid connections that not made to/come from same origin (aka. `CORS`)
- Visit `www.facebook.com` or just `facebook.com` with normal desktop layout, as normal of using on desktop computer or laptop.
  > Simple than two above, only things to detect is **if user connect from mobile data or not?** if yes, then get redirected to `web.facebook.com`.

I don't know how to write and expect to be work? but at the first step, I'm use Fennec F-Droid (F-Droid version of Firefox that cut `Google Play` dependencies) and set some value in `about:config` to override specific user agent for 3 sites above, but...
- Even you can visit `touch.facebook.com` directly, the `AJAX`/`XHR` is still the main problem.
- I have no idea about how to navigate to `www.facebook.com` without getting redirect to `web.facebook.com`, still have no solution, yet.

**Comment by DandelionSprout** [comment]
This is certainly a lot of situations for me to look into, which I'll do once I'll take from a break from an ongoing PaintNET translation to Norwegian that I've begun to work on.

**Comment by DandelionSprout** [comment]
Okay, are you able to download https://raw.githubusercontent.com/DandelionSprout/adfilt/master/Sandbox/Facebook%20phone%20test.json to your computer and to import it into Redirector? (Try to have only one of the two entries turned on at once.)

If so, do these entries work as you had hoped for?

**Comment by DandelionSprout** [comment]
The syntax that is used by Redirector is simple enough to learn.

In Example URL you type one of the pages that you want to redirect.
In Include Pattern you can type what URL pattern you want to redirect.
In Redirect To you can type what pattern you want to redirect to.

The biggest thing to note, is that *Include Pattern* supports wildcard asterisks, but that *Redirect To* uses dollarsign placeholders to figure out which wildcard it shall imitate. E.g., $3 in *Redirect To*, imitates the 3rd wildcard in *Include Pattern*.

**Comment by kowith337** [comment]
Sorry, but I've view the code and seems likely I'm not describe clear enough about what i should be.

The proper goal is
- Reflect redirect to `h.facebook.com` that done by Facebook to trying go to `mobile.facebook.com`, rather than stay on `mbasic.facebook.com` as it should be
- On the `touch.facebook.com` reflect redirect any `XHR` connections that point to `mobile.facebook.com`, rather than stay on the same origin
- Redirect `web.facebook.com` back to `www.facebook.com`

However, I cannot guarantee about possibility of loop or deadlock because Facebook try redirecting as their system wants to, while Redirector trying to revert back with custom rules to return proper connections as it should be.

**Comment by DandelionSprout** [comment]
In that case I'm genuinely unsure how to resolve this, since I struggle to reproduce these behaviours on my end, and because Redirector can for instance not deal with API elements.

I could attempt to update the test file according to what I think you're struggling with, but by now I think it'd be easier if you became able to create your own entries in *Redirect*

**Comment by kowith337** [comment]
OK, I've tried and seems like it doesn't work... :(
You're right about it cannot deal against their API, but on a little bright side, it may solving something, e.g. QR image of loyalty card is loaded from `web.facebook.com` and I've blocked connection of that subdomain (done by [this list](https://github.com/kowith337/PersonalFilterListCollection/raw/master/filterlist/FacebookZuckestionKiller.txt))

But I will thank you for you help, even it doesn't work.

Any way, I'm also thank you about bringing EasyList Thailand to submit into uBlock regional list.
